[UI] Deep Link search in /secrets

Manish-Giri opened the pull request:
<!-- Did you remember to add a changelog snippet? See
     https://github.com/taskcluster/taskcluster/blob/master/dev-docs/best-practices/changelog.md

     If this is related to a Bugzilla bug, please begin your title with [Bug XXXXX]
     and update this link.  Otherwise, just remove it from your PR comment.  -->

<!-- If this is related to a GitHub Bug/Issue, Please write Issue Number after # in the next line. Otherwise, just remove it from your PR comment. -->

@helfi92 I noticed the intended behavior for `/secrets` already exists in `/hooks`, and can be seen here (as an example) - https://firefox-ci-tc.services.mozilla.com/hooks?search=project-

So I figured it might be worth a try to replicate the functionality already present in `ListHooks/index.jsx`. 

This change uses the `qs` package to parse out the search term from the query string, and pushes it onto `history`. Thus, once you do the following steps -
1. Visit `/secrets` and search for an item, example `project/`
2. Click on a secret
3. Go back

The browser URL now contains the search term you searched for -

![Screenshot 2020-04-13 23 34 30](https://user-images.githubusercontent.com/11348778/79203699-64d97600-7e09-11ea-9256-cac1e6aae77c.png)

whereas it is currently just `/secrets` after you go back -

![Screenshot 2020-04-13 22 08 03](https://user-images.githubusercontent.com/11348778/79203773-83d80800-7e09-11ea-8b87-82b9f71cdc90.png)

However, the page still lists all the available results at `/secrets`. 

I wasn't quite sure about why this behavior at `/secrets` is different from `/hooks`, so I figured it would be nice to get your thoughts on this, if possible, before I investigate further. I wonder if the `refetch()` in the Submit Handler from [here](https://github.com/taskcluster/taskcluster/blob/master/ui/src/views/Secrets/ViewSecrets/index.jsx#L46-L53) makes any difference. However, the [docs](https://www.apollographql.com/docs/react/data/queries/#refetching) from Apollo GraphQL mention this just refreshes the query results in response to a user action, and as far as I can see, the `filter` still lists the correct search string -

Would be really great if I could some thoughts/tips about the next steps!

![Screenshot 2020-04-13 23 33 02](https://user-images.githubusercontent.com/11348778/79204796-1a58f900-7e0b-11ea-8c40-13231fee7db5.png)

Github Bug/Issue: Fixes #2123

helfi92 commented on ui/src/views/Secrets/ViewSecrets/index.jsx:
```diff
@@ -36,6 +37,18 @@ export default class ViewSecrets extends Component {
+    }
+
+    return null;
+  }
```
I just created https://github.com/taskcluster/taskcluster/pull/2660/commits/bd9db1a9905a204a816379468d401394e6ef740c to move away from this kind of logic 😅 See how there's a simpler way to do this without relying on internal state.

helfi92 commented on ui/src/views/Secrets/ViewSecrets/index.jsx:
```diff
@@ -51,6 +64,10 @@ export default class ViewSecrets extends Component {
     });
+
+    this.props.history.push(
+      `/secrets${secretSearch ? `?search=${secretSearch}` : ''}`
```
Let's encode the search term before pushing it to history similar to how it's done in https://github.com/taskcluster/taskcluster/pull/2660/commits/bd9db1a9905a204a816379468d401394e6ef740c.

helfi92 replied:
On a second thought, there's no need to encode if we use the `stringify` function from the `qs` package. https://github.com/taskcluster/taskcluster/blob/978b8dc93e54d5ec34cfbd06cad6c88796f53f5f/ui/src/views/Hooks/ListHooks/index.jsx#L36-L45

Manish-Giri commented:
@helfi92 Sorry for the delay. I have been trying to explore React Router, especially `history` for a while now, to figure out what the possible cause might be, for the incorrect behavior of the back button after a search, but couldn't come up with a concrete reason.

Essentially, in the latest commit, I made the following changes -
1. Replaced internal state with the search being derived from `window.location.search`, following the approach in your recent commit, as you pointed out.
2. Also updated the PageChangeHandler accordingly, in addition to the Submit Handler, to not rely on `state`.
3. Added a `defaultValue` prop to the `Search` component in the `render()` function (I was testing to see if this made a difference...).

But, even though the search results correctly display the search term, this behavior doesn't persist after pressing the back button (displayed in screenshots below) -

1. Search for `cia/` -

![Screenshot 2020-04-26 05 12 34](https://user-images.githubusercontent.com/11348778/80303210-cb9d4e80-877c-11ea-8a1f-5b1f15c1e8af.png)

2. Click on one of the search results and go back -

![Screenshot 2020-04-26 05 13 02](https://user-images.githubusercontent.com/11348778/80303249-02736480-877d-11ea-88c4-9cf5ef4b5039.png)

helfi92 commented:
@Manish-Giri Let me know when you want me to review :-)

Manish-Giri commented:
@helfi92 Thank You! I think I understand your suggested changes. I’ll fix this right away and re-submit.

helfi92 commented:
Awesome! Thanks @Manish-Giri :-)

Manish-Giri commented:
@helfi92 Sorry for the delay, once again :/

The search works perfectly now - only secrets that include the search term that was searched for originally, are now shown, when you click on the searched secret, and go back. Here is a screenshot for search of`project/` -  only secrets with `project/` are now shown when you go back -

![Screenshot 2020-05-18 04 49 55](https://user-images.githubusercontent.com/11348778/82197665-41529100-98c9-11ea-9df7-1951a2a116e2.png)

However, while rebasing with the latest `master`, I noticed there was some work done on [this file](https://github.com/taskcluster/taskcluster/pull/2617/files) as part of #2617, which included maintaining a `state` in the `ViewSecrets` component. Since we were trying to eliminate the need for maintaining a state in the component, I removed `state`, while keeping other changes from that PR. 
The only thing I was a bit confused about is this line here - 
https://github.com/taskcluster/taskcluster/blob/278a92188b830699b7a75be0badeb96106f5d490/ui/src/views/Secrets/ViewSecrets/index.jsx#L101

which, I think, attempts to destructure `searchTerm` from `state` and then pass it on as a prop to the `SecretsTable` component -
https://github.com/taskcluster/taskcluster/blob/278a92188b830699b7a75be0badeb96106f5d490/ui/src/views/Secrets/ViewSecrets/index.jsx#L119

However, `state` only contains a `secretSearch` property, as far as I could tell, so not sure what's going on there. I assume this is meant to capture the search term, if so, I simply used the existing search term we already obtained from `qs.parse` and then pass it on as the `searchTerm` prop. 

```
const query = qs.parse(this.props.location.search.slice(1));
const secretSearch = query.search;
```
If that is not the case, I can make changes accordingly. 

Thank You!

Manish-Giri commented:
@helfi92 I rebased it with the latest `master`. I didn't know if the PR needed a changelog, so I didn't add it in, looks like the missing changelog might be one of the errors. If you think it needs the changelog, please let me know an appropriate changelog message and I'll add it in. Thank You!

Manish-Giri commented:
@helfi92 Thank You so much for all your help with this PR. Should I now start looking at #2124 (fix search in `/worker-manager`), since the underlying issue is the same?

As a side note, I wonder if it would be possible to include me in the [Taskcluster contributors list on the repo](https://github.com/taskcluster/taskcluster#contributors). Sorry if this is a restricted list, but I figured it wouldn't hurt to ask, it would just mean a lot to me to have been able to contribute to Taskcluster 'officially'!

helfi92 commented:
> @helfi92 Thank You so much for all your help with this PR. Should I now start looking at #2124 (fix search in `/worker-manager`), since the underlying issue is the same?

That would be so awesome. I'll go ahead and assign you to the issue now. Thanks for helping as always!

> 
> As a side note, I wonder if it would be possible to include me in the [Taskcluster contributors list on the repo](https://github.com/taskcluster/taskcluster#contributors). Sorry if this is a restricted list, but I figured it wouldn't hurt to ask, it would just mean a lot to me to have been able to contribute to Taskcluster 'officially'!

Of course! You should definitely be on that list. I'm on it :)

helfi92 commented:
@all-contributors please add @Manish-Giri for code.
